wire-ios-data-model/Source/ConversationSearch/MutableConversationContainer.swift

```diff
@@ -0,0 +1,27 @@
+//
+// Wire
+// Copyright (C) 2024 Wire Swiss GmbH
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful,
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE. See the
+// GNU General Public License for more details.
+//
+// You should have received a copy of the GNU General Public License
+// along with this program. If not, see http://www.gnu.org/licenses/.
+//
+
+/// A mutable container of conversations used by the `ConversationSearchUseCase`.
+/// This can be interpreted as a section in a table where a row is a conversation.
+public protocol MutableConversationContainer {
```

Comment:
It can be anything as long as it satisfies the requirements (`conversations` getter and `removeConversation(at:)` function). 

In the navigation overhaul branch the type `ConversationListViewModel.Section` conforms to this protocol.

Reply:
Take a look into `ConversationListViewModel.Section`. It is used as a conversation-grouping model for the conversation list, but has the `fileprivate` access modifier and is not implemented in a clean way, in my opinion.

So I saw three options:
1. Implementing the filter algorithm (`FilterConversationsUseCase`) in Wire-iOS on top of the UI code using the actual types declared there (`ConversationListViewModel` etc.): I actually don't consider it an option.
2. Abstracting the search algorithm using simple collection types (like Array): Then unfortunately a double mapping is required, first from the existing data structure (`ConversationListViewModel.Section`) to an array and after the filtering, mapping the filtered array back to the custom data structure.
3. Declaring the generic `MutableConversationContainer` which allows for getting the current state and removing (= filtering out) items. This is the cleanest approach in my opinion.

Reply:
This is the original PR, btw:
https://github.com/wireapp/wire-ios/pull/1605

Reply:
It seems that without refactoring also the view layer here we are limited in nice solutions and I agree that the one you have chosen is the cleanest option at the moment.

Ideally however I would imagine that the view model requests a simple grouped list of conversations via a use case, passing in an optional filter. I think the filter implementation shouldn't be constrained by the implementation of the view, but rather the view needs to work with a well designed interface provided to it.

For instance:

```swift
let conversationsByCategory: [ConversationCategory: [Conversation]] = usecase.invoke(filter: .query("foo"))
```

Reply:
Yes, this would be nice.
I vote for rewriting the conversation list in SwiftUI eventually.